versioned/storage/store/src/main/java/org/projectnessie/versioned/storage/versionstore/VersionStoreImpl.java

```diff
@@ -251,16 +258,29 @@ public ReferenceAssignedResult assign(NamedRef namedRef, Hash expectedHash, Hash
     }
 
     try {
-      CommitObj head = commitLogic(persist).headCommit(expected);
-      Hash currentCommitId = head != null ? objIdToHash(head.id()) : NO_ANCESTOR;
+      Hash currentCommitId;
+      try {
+        CommitObj head = commitLogic(persist).headCommit(expected);
+        currentCommitId = head != null ? objIdToHash(head.id()) : NO_ANCESTOR;
+      } catch (ObjNotFoundException e) {
+        // Special case, when the commit object does not exist. This situation can happen in a
+        // multi-zone/multi-region database setup, when the "primary write target cluster" fails and
+        // the replica cluster(s) had no chance to receive the commit object but the change to the
+        // reference. In case the zone/region is permanently switched (think: data center flooded /
+        // burned down), this code path allows to assign the branch to another commit object, even
+        // if the current HEAD's commit object does not exist.
+        currentCommitId = expectedHash;
+      }
+
       verifyExpectedHash(currentCommitId, namedRef, expectedHash);
       expected =
           reference(
               refName,
               hashToObjId(expectedHash),
               false,
               expected.createdAtMicros(),
-              expected.extendedInfoObj());
+              expected.extendedInfoObj(),
+              expected.previousPointers());
```

Comment:
So we're collecting the history of the named reference itself in `previousPointers()`. If a reference is reassigned, a few commits are made and then there's storage failure. Some of the new commits may not be replicated. With current `previousPointers()` we'd be able to recover the reference to its previous location in the total tree of commits. However, I wonder whether users may want to recover to the last "good" predecessor of the most recent reference HEAD (which may be unrelated to the older entries in `previousPointers()`. WDYT?

Reply:
That's why there's the functionality to optionally retrieve the commit-log from the current HEAD

Reply:
but if HEAD is not replicated, the only option is to go back into reference history, which may take us back to a different place in the commit tree. With ref assignments the reference's history will not always match the commit lineage of the most recent HEAD.